### PR TITLE
fix(pnp): always initialize the ESM loader when enabled

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

Dynamic imports doesn't work in commonjs files when the entrypoint is also a commonjs module.

Fixes https://github.com/yarnpkg/berry/issues/3671
Fixes https://github.com/yarnpkg/berry/issues/3687
Fixes PnP support in the Babel repo (cc @nicolo-ribaudo)
Fixes (part of) the failing Angular e2e test

**How did you fix it?**

- Update the ESM loader to handle extensionless files (workaround for https://github.com/nodejs/node/issues/33226), this allows removing the patch to `require('module').runMain`. Removing that patch lets Node initialize the loader regardless of the module type
- Update `process.mainModule` when the commonjs module is already in the cache when loading the main module

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/acceptance-tests/pkg-tests-specs/sources/pnp-esm.test.ts [hunk 1]
```diff
@@ -141,57 +141,6 @@ describe(`Plug'n'Play - ESM`, () => {
...
-        await expect(run(`node`, `./index.js`)).resolves.toMatchObject({
-          code: 0,
-          stdout: `42\n`,
-        });
-      },
-    ),
-  );
-
-  test(
-    `it should always set default as module.exports when importing a commonjs file`,
```
Comment (arcanis) [inline packages/acceptance-tests/pkg-tests-specs/sources/pnp-esm.test.ts:170]: Quick note: it's worth adding a (in-PR) comment when tests are removed 😃

Comment (merceyz) [reply packages/acceptance-tests/pkg-tests-specs/sources/pnp-esm.test.ts:170]: I wrote it on the commit that removed them

**Comment by timmywil** [comment]
Could this be released to canary?

**Comment by merceyz** [comment]
Sure, running it now https://github.com/yarnpkg/berry/actions/runs/1468300994

Alternatively you can get it from sources
```sh
yarn set version from sources
```

**Comment by teohhanhui** [comment]
Looks like that has failed: https://github.com/yarnpkg/berry/runs/4229011500?check_suite_focus=true#step:7:1257

**Comment by merceyz** [comment]
The publish to the registry failed but `set version...` doesn't use the registry so it doesn't matter.
```sh
yarn set version canary
```

**Comment by kherock** [comment]
The extensionless file workaround for https://github.com/nodejs/node/issues/33226 in this PR doesn't work when I'm attempting to load an entrypoint with an arbitrary file extension that's intended to be parsed by a `require` hook like `babel-register`. I have a local package that loads an entrypoint with a `.ts` extension that should normally load as CJS.

Are there any gotchas to renaming the ``` ``:``` and `` `.js`:`` cases to the `default:` case? I expect `.ts` to load as CJS when the package doesn't declare `"type": "module"`. This assumption might change once `.mts` exists, but defaulting to the package type feels like the most correct behavior.

**Comment by merceyz** [comment]
Could you open a seperate issue to track that?

> Are there any gotchas to renaming the ``: and `.js`: cases to the default: case?

We don't want to deviate from how Node behaves more than we have to so we could only do that for the entrypoint

**Comment by tavoyne** [comment]
I'm experiencing an analogous issue that this PR doesn't seem to be fixing.

I'm trying to use the `@mdx-js/loader@2.0.0-rc.2` webpack loader. In the [package's entry point](https://github.com/mdx-js/mdx/blob/main/packages/loader/index.cjs) (it's a `.cjs` file), there's a dynamic import of an ES module that causes the following error:
```
node:internal/errors:464
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/blackjelly/code/blog-3/.yarn/__virtual__/@mdx-js-loader-virtual-a537165eb1/0/cache/@mdx-js-loader-npm-2.0.0-rc.2-eb4487e27d-5637b8d92c.zip/node_modules/@mdx-js/loader/lib/index.js' imported from /Users/blackjelly/code/blog-3/.yarn/__virtual__/@mdx-js-loader-virtual-a537165eb1/0/cache/@mdx-js-loader-npm-2.0.0-rc.2-eb4487e27d-5637b8d92c.zip/node_modules/@mdx-js/loader/index.cjs
```

Which setting Yarn's version from sources or using canary doesn't fix.

**Comment by merceyz** [comment]
Please open a bug report with a reproduction

**Pull request merged**